### PR TITLE
Colony rotation and drift again

### LINKED ISSUE #2511: Binding to a cell with pili results in ghost drifting
**Describe the bug**

When binding with a cell that has pili the ghost drifting bug occurs again.

| Player Cell | Member Lvl 1 | Member Lvl 2 | Result |
|-------------|--------------|--------------|--------|
| No Pili     | No Pili      | No Pili      | Fine   |
| No Pili     | No Pili      | Pili         | Fine   |
| No Pili     | Pili         | No Pili      | Bugged |
| No Pili     | Pili         | Pili         | Bugged |
| Pili        | No Pili      | No Pili      | Fine   |
| Pili        | No Pili      | Pili         | Fine   |
| Pili        | Pili         | No Pili      | Bugged |
| Pili        | Pili         | Pili         | Bugged |

**Additional context**

Save file to test with: [auto_save_1.zip](https://github.com/Revolutionary-Games/Thrive/files/7022280/auto_save_1.zip)

### LINKED ISSUE #2512: Level3 colony members shape rotations are incorrect
**Describe the bug**

Binding the 2nd colony member to a 3rd in the chain causes the 3rd not to have correct shape alignment.

**To Reproduce**
Bind to 1 cell.
Bind to a 2nd cell through the cell you just attached.
Bind to a 3rd cell through the 2nd cell.

**Expected behavior**
Shapes sould be aligned to the cell

**Screenshots**
image

### PR DESCRIPTION
**Pull request by Kemikal1**
**Brief Description of What This PR Does**

Fixes ghost drift and rotation in colonies for good. There shouldnt be anymore bugs for the colony, except maybe some random rare detaching.

**Related Issues**
Fixes #2511 
Fixes #2512 

**Progress Checklist**

Note: before starting this checklist the issue should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by FuJa0815** [comment]
> 
> 
> Seems to that the ghost drift and rotation problems were fixed. However, I was unable to go into the editor even though I heard the sound indicating that I was able to. Along with that I was unable to replicate myself through cheats. Plus for good measure, (although, I'm not sure if it's related) sometimes the toxic pink mist would appear and disappear at random times.
> [Wont let me go into editor or let me duplicate.zip](https://github.com/Revolutionary-Games/Thrive/files/7068571/Wont.let.me.go.into.editor.or.let.me.duplicate.zip)
> [Pink Mist Vanishes AND Cant go into editor or duplicate.zip](https://github.com/Revolutionary-Games/Thrive/files/7068572/Pink.Mist.Vanishes.AND.Cant.go.into.editor.or.duplicate.zip)
> 
> 

You cannot enter the editor while in a colony. We probbably should make a tooltip about that. I think that the sound playing is not a bug because it indicates the player that you may leave the colony now and go to the editor.
The toxic pink bug is unrelated and also on main. Is there an open issue about this?

**Comment by Kemikal1** [comment]
Thank god,was about to test it,and i was already thinking about what could be the problem

**Comment by hhyyrylainen** [comment]
Perhaps something like how the current tooltip reads:


The editor button (nor the sound related to it) should *not* become enabled while in a colony. That is a bug (too lazy to look up which issue number that is)

File: src/microbe_stage/PlacedOrganelle.cs [hunk 1]
```diff
@@ -409,9 +409,12 @@ public void ReParentShapes(Microbe to, Vector3 offset, Vector3 masterRotation, V
             {
                 // TODO: quaternion usage would be good here
                 // https://github.com/Revolutionary-Games/Thrive/issues/2504
-                shapePosition = shapePosition.Rotated(Vector3.Up, parentRotation.y);
-                if (ParentMicrobe.ColonyParent != ParentMicrobe.Colony.Master)
-                    shapePosition = shapePosition.Rotated(Vector3.Up, masterRotation.y);
+                var parent = ParentMicrobe;
+                while (parent != ParentMicrobe.Colony.Master)
+                {
+                    shapePosition = shapePosition.Rotated(Vector3.Up, parent.Rotation.y);
```
[the remark this answers is not in the document]
Comment (Kemikal1) [reply src/microbe_stage/PlacedOrganelle.cs:415]: Why quaternion?Wouldnt i need Vector3 for a transform?

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/PlacedOrganelle.cs:415]: Quaternions are rotations, so you could build the total rotation as a quaternion and then rotate the vector by that quaternion.

[the remark this answers is not in the document]
Comment (Kemikal1) [reply src/microbe_stage/PlacedOrganelle.cs:415]: How?

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/PlacedOrganelle.cs:415]: Get the quaternion from the parent, and multiply those together until there isn't a parent anymore. Then after that rotate the vector by the quaternion. Basically very similar to what this code does, except it doesn't assume that only y axis converted radian rotation is useful.

[the remark this answers is not in the document]
Comment (Kemikal1) [reply src/microbe_stage/PlacedOrganelle.cs:415]: Is there a method to get the quat of the parent?

> rotate the vector by the quaternion.
How?this is what im asking.

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/PlacedOrganelle.cs:415]: In Godot rotations is a Basis type, out of which you can separate out the rotation and scaling.

By digging through auto-complete I found this method in Quat:

```
    /// <summary>
    /// Returns a vector transformed (multiplied) by this quaternion.
    /// </summary>
    /// <param name="v">A vector to transform.</param>
    /// <returns>The transformed vector.</returns>
    /// <footer><a href="https://www.google.com/search?q=Godot.Quat.Xform">`Quat.Xform` on google.com</a></footer>
    public Vector3 Xform(Vector3 v)
```

[the remark this answers is not in the document]
Comment (Kemikal1) [reply src/microbe_stage/PlacedOrganelle.cs]: Thing is it the verbs here cause confusion.Transform ,apply,multiply..could mean different things.

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/microbe_stage/PlacedOrganelle.cs]: If you check all the different descriptions for the various operations, I think it is pretty clear. At least with my level of experience with 3D math.

**Comment by sirdanforth** [comment]
> > Seems to that the ghost drift and rotation problems were fixed. However, I was unable to go into the editor even though I heard the sound indicating that I was able to. Along with that I was unable to replicate myself through cheats. Plus for good measure, (although, I'm not sure if it's related) sometimes the toxic pink mist would appear and disappear at random times.
> > [Wont let me go into editor or let me duplicate.zip](https://github.com/Revolutionary-Games/Thrive/files/7068571/Wont.let.me.go.into.editor.or.let.me.duplicate.zip)
> > [Pink Mist Vanishes AND Cant go into editor or duplicate.zip](https://github.com/Revolutionary-Games/Thrive/files/7068572/Pink.Mist.Vanishes.AND.Cant.go.into.editor.or.duplicate.zip)
> 
> You cannot enter the editor while in a colony. We probbably should make a tooltip about that. I think that the sound playing is not a bug because it indicates the player that you may leave the colony now and go to the editor.
> The toxic pink bug is unrelated and also on main. Is there an open issue about this?

Ah! That makes sense lol I think that if it was listed along side the engulf and bind buttons it would be clearer. Still unable to replicate myself on the cheat menu though.

**Comment by Kemikal1** [comment]
> @Nunez2196 posted a bunch of links on the programming channel on discord. If you don't use quaternions for rotation, you can get into gimbal lock -> it's not possible to represent all possible rotations without using quaternions -> we must use quaternions without losing information, otherwise things are not done properly.



> @Nunez2196 posted a bunch of links on the programming channel on discord. If you don't use quaternions for rotation, you can get into gimbal lock -> it's not possible to represent all possible rotations without using quaternions -> we must use quaternions without losing information, otherwise things are not done properly.

Gimbal lock refers to apllying and calculating the rotation not representing it.

**Comment by hhyyrylainen** [comment]
>  Gimbal lock refers to apllying and calculating the rotation not representing it.

Isn't it the same thing? You can't represent a quaternion with 3 euler angles. You've already lost data. This is because a quaternion is a vector and an angle of rotation along that vector. Meaning that 3 angles can't represent that, because certain combinations of angles result in gimbal lock.

I'd really like someone more experienced with 3D math to step in to this discussion...

**Comment by Kemikal1** [comment]
> `Like you lose the shortest path to that rotation but you regain it after converting again to Quat,
Quat ->Euler ->Quat loses no relevant information regarding `rotation.``

**Comment by hhyyrylainen** [comment]
In our case it doesn't because we have one axis locked, so 3 angles can represent all possible rotations in the microbe stage. But my whole point is that this code should be made general ie. use Quats to be able to handle any rotation (and don't cut off information in the middle).

**Comment by Kasterisk** [comment]
> Still unable to replicate myself on the cheat menu though.

It's this issue #2519, only occurs after exiting from the initial editor session.

**Comment by hhyyrylainen** [comment]
I suppose this is waiting for moving to quaternions as the next step, so I'll remove this from my review backlog for now.

**Comment by Kemikal1** [comment]
Move what to quaternions?If you mean the rest of the rotation calculations i think that needs another PR , its an entirely different goal from this PR.

**Comment by hhyyrylainen** [comment]
I meant the calculations that this PR modified. Though, it's possible that I misremember and this PR doesn't actually modify any of the math...

**Comment by Kemikal1** [comment]
This PR midified to quaternions,what are you talking about?

**Comment by hhyyrylainen** [comment]
I see.
I guess then I'd like to see someone else first review this (maybe @FuJa0815 who has an existing request for changes still on this PR) before I take the last look.

**Comment by FuJa0815** [comment]
Pili is not damaging enemies when part of a colony children.
[nodamage.thrivesave.gz](https://github.com/Revolutionary-Games/Thrive/files/7154729/nodamage.thrivesave.gz)

**Comment by FuJa0815** [comment]
> If i bind the unbind with the cells

I don't understand.

If you load the savegame and you don't unbind or bind, instead you cheat more speed and touch the enemy cell with the pili of the colony children

**Comment by Kemikal1** [comment]
After i load the save, if i unbind ,then bind to the same cells, the pili works i mean.

**Comment by Kemikal1** [comment]
@FuJa0815 Can you replicate the issue on another save and with the current version?

**Comment by FuJa0815** [comment]
Huh, I am not able to replicate this? Maybe I was testing an old version, but I doubt it... Maybe this was a rare bug?

**Review by FuJa0815 — APPROVED** [review]
should be fine if this bug really does not occur anymore. Playtested it for 7min.

**Comment by hhyyrylainen** [comment]
I merged master into this, which had that one MicrobeColony constructor changing PR in it, so re-testing this with that included would be nice.

**Comment by Kemikal1** [comment]
The bug happens only on load,and kind of often, extends to microbes that attach to colony after load too. I think the only way of fixing this is just rebuilding the colony at load.

**Comment by Kemikal1** [comment]
Before fixing formats i would like people to test and review my code, it changes kinda much so i dont want to lose time with useless things like that.

File: src/microbe_stage/Microbe.cs [hunk 2]
```diff
@@ -1412,6 +1459,37 @@ public override void _IntegrateForces(PhysicsDirectBodyState physicsState)
...
+    public void RemakeColony()
+    {
+        var col = new List<Microbe>(ColonyChildren);
+        var sem = 0;
+
+        // Check if the pili of colony members have been formed
+        foreach (var member in col)
+        {
+            foreach (var organelle in member.organelles)
+                if (organelle.Definition.Name == "Predatory Pilus" && member.pilusPhysicsShapes.Count == 0)
```
Comment (FuJa0815) [inline src/microbe_stage/Microbe.cs:1474]: https://github.com/Revolutionary-Games/Thrive/pull/2379#discussion_r667486970

Comment (Kemikal1) [reply src/microbe_stage/Microbe.cs:1474]: Lol yea, ill fix it.

Comment (Kemikal1) [reply src/microbe_stage/Microbe.cs:1474]: Is that the only  code problem u saw?

**Comment by FuJa0815** [comment]
What is the current state of this PR?

**Comment by Kemikal1** [comment]
Its done , just some formatting things that i said, i will do after someone playtests it and gives me some kind of green light of approval.

**Comment by FuJa0815** [comment]
Short ingame testing looked very good, haven't had time to check out the code yet

**Pull request merged**